### DISCUSSION TITLE
incoming() in tokio 1.0?

### DISCUSSION THREAD

**Discussion by Roguelazer** [tokio-rs/tokio#3380] [General] [answered]
The `incoming()` method on Listeners was a very convenient way to abstract across different listeners (e.g., `TcpListener` and `UnixListener`) until we have some form of HKT. I see it was removed in #2928; does anyone know if there are plans to put it back (either in `tokio` proper or in `tokio-stream` as an extension trait)? Some of my code will be quite tricky to port from tokio 0.2 to tokio 1.0 (or, at least, have a lot more duplication) until it returns...

**Comment by taiki-e** [comment, marked as answer]
You can use tokio_stream::wrappers::{[TcpListenerStream](https://github.com/tokio-rs/tokio/blob/e42317b072c5240fe190c0a8708cd24bb4a4eb08/tokio-stream/src/wrappers/tcp_listener.rs
), [UnixListenerStream](https://github.com/tokio-rs/tokio/blob/e42317b072c5240fe190c0a8708cd24bb4a4eb08/tokio-stream/src/wrappers/unix_listener.rs)}. Both are the same as 0.2's `incoming`.

**Reply by Darksonn** [reply]
Note that the `tokio-stream`'s docs for the wrappers module is currently generated incorrectly, so you can't see them in the documentation unless you build it locally. Note also that you should enable the appropriate feature of tokio-stream.

**Reply by taiki-e** [reply]
Hmm, that's true. Probably worth creating a new release...

**Reply by Roguelazer** [reply]
I'm glad this functionality is somewhere. I never would've known if you hadn't mentioned it, though, and would definitely endorse fixing the documentation.
